File: src/python/vrp/instance_reader.py
#  Copyright (c) 2025 Laboratory for Combinatorial Optimization in Real-time Environment.
#  All rights reserved.

from vrp.instance import Instance
import json


class InstanceReader:
    def __init__(self, file_path: str):
        self.file_path_ = file_path
        self.solutions = None

    def read(self) -> Instance:
        print("InstanceReader::read()")
        nb_vehicles = 0
        capacity = 0

        if "_" in self.file_path_:
            path = self.file_path_.split("/")
            instance_name = path[-1].split(".")[0]
            if "RC" in instance_name:
                Itype = "RC"
            elif "R" in instance_name:
                Itype = "R"
            elif "C" in instance_name:
                Itype = "C"
            self.file_path_ = "/".join(path[:-1]) + f"/generated/{Itype}/{path[-1]}"

        print(f"file_path_={self.file_path_}")

        with open(self.file_path_, "r") as f:
            # First line contains the name of the instance
            instance_name = f.readline().strip()

            # Skip lines 2 to 4
            f.readline()
            f.readline()
            f.readline()

            # Line 5 contains the number of vehicles and the vehicle capacity
            nb_vehicles, capacity = map(int, f.readline().split())

            instance = Instance(nb_vehicles, capacity, instance_name)

            # Skip lines 6 to 9
            f.readline()
            f.readline()
            f.readline()
            f.readline()

            # The remaining lines contain information about customers
            for line in f:
                parts = line.strip().split()
                if not parts:
                    continue

                id = int(parts[0])
                pos_x = float(parts[1])
                pos_y = float(parts[2])
                demand = int(parts[3])
                ready_time = int(parts[4])
                due_time = int(parts[5])
                service_time = int(parts[6])

                depot = id == 0
                instance.add_customer(
                    id, pos_x, pos_y, demand, ready_time, due_time, service_time, depot
                )

        print(f"nb_customers: {len(instance.get_customers_by_id())}")
        return instance

    def read_duals(self, duals_file_path: str) -> dict[int, float]:
        print("read_duals")
        dual_by_var_id: dict[int, float] = {}

        with open(duals_file_path, "r") as f:
            for line in f:
                parts = line.strip().split()
                if not parts:
                    continue
                id = int(parts[0])
                dual_value = float(parts[1])
                dual_by_var_id[id] = dual_value

        return dual_by_var_id
    
    def read_dual_optimal(self, instance_name:str) -> dict[int, float]:
        if self.solutions is None:
            with open("../../instances/duals/duaux_optimaux.json", "r") as f:
                self.solutions = json.load(f)
        dual_by_id_str: dict[str, float] = self.solutions[instance_name]
        dual_by_id:dict[int, float] = {}

        for i in dual_by_id_str:
            dual_by_id[int(i)] = dual_by_id_str[i]

        return dual_by_id

File: src/vrp/instance_reader.cpp
// Copyright (c) 2025 Laboratory for Combinatorial Optimization in Real-time Environment.
// All rights reserved.

#include "instance_reader.hpp"

#include <fstream>
#include <iostream>

InstanceReader::InstanceReader(std::string file_path) : file_path_(std::move(file_path)) {}

Instance InstanceReader::read() const {
    std::cout << "InstanceReader::read()\n";

    int nb_vehicles = 0;
    int capacity = 0;
    std::string instance_name;

    std::cout << "file_path_=" << file_path_ << std::endl;

    std::ifstream file(file_path_);

    // First line contains the name of the instance.
    std::getline(file, instance_name);

    // Skip lines 2 to 4.
    std::string line;
    std::getline(file, line);
    std::getline(file, line);
    std::getline(file, line);

    // Line 5 contains the number of vehicles and the vehicle capacity.
    file >> nb_vehicles >> capacity;

    Instance instance(nb_vehicles, capacity, instance_name);

    // Skip lines 6 to 9.
    std::getline(file, line);
    std::getline(file, line);
    std::getline(file, line);
    std::getline(file, line);

    // The remaining lines information about customers.
    int customer_id = 0;
    double pos_x = 0;
    double pos_y = 0;
    int demand = 0;
    int ready_time = 0;
    int due_time = 0;
    int service_time = 0;

    while (file >> customer_id >> pos_x >> pos_y >> demand >> ready_time >> due_time >>
           service_time) {
        bool depot = false;
        if (customer_id == 0) {
            depot = true;
        }

        instance.add_customer(customer_id,
                              pos_x,
                              pos_y,
                              demand,
                              ready_time,
                              due_time,
                              service_time,
                              depot);
    }

    std::cout << "nb_customers: " << instance.get_customers_by_id().size() << std::endl;

    return instance;
}

std::map<size_t, double> InstanceReader::read_duals(const std::string& duals_file_path) {
    std::cout << __FUNCTION__ << std::endl;

    std::map<size_t, double> dual_by_var_id;

    std::ifstream file(duals_file_path);

    int var_id = 0;
    double dual_value = 0.0;

    while (file >> var_id >> dual_value) {
        dual_by_var_id.emplace(var_id, dual_value);
    }

    return dual_by_var_id;
}
